src/Crdt/SnapshotWorker.cs

```diff
@@ -0,0 +1,180 @@
+/// </summary>
+public class SnapshotWorker
+{
+    private readonly IReadOnlyDictionary<Guid, SimpleSnapshot>? _snapshots;
+    private readonly CrdtRepository _crdtRepository;
+    private readonly Dictionary<Guid, ObjectSnapshot> _pendingSnapshots  = [];
+    private readonly List<ObjectSnapshot> _newIntermediateSnapshots = [];
+
+    private SnapshotWorker(Dictionary<Guid, ObjectSnapshot> snapshots, CrdtRepository crdtRepository)
+    {
+        _pendingSnapshots = snapshots;
+        var oldestSnapshot = snapshots.Values.MinBy(s => s.Commit.CompareKey);
+        _crdtRepository = crdtRepository;
+    }
+
+    public static async Task<Dictionary<Guid, ObjectSnapshot>> ApplyCommitsToSnapshots(Dictionary<Guid, ObjectSnapshot> snapshots,
+        CrdtRepository crdtRepository,
+        ICollection<Commit> commits)
+    {
+        //we need to pass in the snapshots because we expect it to be modified, this is intended.
+        //if the constructor makes a copy in the future this will need to be updated
+        await new SnapshotWorker(snapshots, crdtRepository).ApplyCommitChanges(commits, false, null);
+        return snapshots;
+    }
+
+    public SnapshotWorker(IReadOnlyDictionary<Guid, SimpleSnapshot> snapshots, CrdtRepository crdtRepository)
+    {
+        _snapshots = snapshots;
+        _crdtRepository = crdtRepository;
+    }
+
+    public async Task UpdateSnapshots(Commit oldestAddedCommit)
+    {
+        var previousCommit = await _crdtRepository.FindPreviousCommit(oldestAddedCommit);
+        var commits = await _crdtRepository.GetCommitsAfter(previousCommit);
+        await ApplyCommitChanges(commits, true, previousCommit?.Hash);
+
+        //intermediate snapshots should be added first, as the last snapshot added for an entity will be used in the projected tables
+        await _crdtRepository.AddIfNew(_newIntermediateSnapshots);
+        await _crdtRepository.AddSnapshots(_pendingSnapshots.Values);
+    }
+
+    public async ValueTask ApplyCommitChanges(ICollection<Commit> commits, bool updateCommitHash, string? previousCommitHash)
+    {
+        var commitIndex = 0;
+        foreach (var commit in commits)
+        {
+            if (updateCommitHash && previousCommitHash is not null)
+            {
+                //we're rewriting history, so we need to update the previous commit hash
+                commit.SetParentHash(previousCommitHash);
+            }
+
+            previousCommitHash = commit.Hash;
+            commitIndex++;
+            foreach (var commitChange in commit.ChangeEntities.OrderBy(c => c.Index))
+            {
+                IObjectBase entity;
+                var snapshot = await GetSnapshot(commitChange.EntityId);
+                var hasBeenApplied = snapshot is not null && (snapshot.CommitId == commit.Id ||
+                                                              snapshot.Commit.HybridDateTime > commit.HybridDateTime);
```

Comment:
yes, the current snapshot is the state of all previously applied commits.

I'm actually wondering if this could be removed now, it might not be needed as hasBeenApplied should always be false I think. I'll need to check that out.

Reply:
I was just able to remove this entierly